conda_index/index/sqlitecache.py

```diff
@@ -484,25 +485,42 @@ def listdir_stat():
 
     def changed_packages(self):
         """
-        Compare 'fs' to 'indexed' state.
+        Compare upstream to 'indexed' state.
 
-        Return packages in 'fs' that are changed or missing compared to 'indexed'.
+        Return packages in upstream that are changed or missing compared to 'indexed'.
         """
-        return self.db.execute(
+        query = self.db.execute(
             """
             WITH
             fs AS
-                ( SELECT path, mtime, size FROM stat WHERE stage = 'fs' ),
+                ( SELECT path, mtime, size, sha256, md5 FROM stat WHERE stage = :upstream_stage ),
             cached AS
-                ( SELECT path, mtime FROM stat WHERE stage = 'indexed' )
+                ( SELECT path, mtime, size, sha256, md5 FROM stat WHERE stage = 'indexed' )
 
-            SELECT fs.path, fs.mtime, fs.size from fs LEFT JOIN cached USING (path)
+            SELECT fs.path, fs.mtime, fs.size, fs.sha256, fs.md5,
+                cached.mtime as cached_mtime, cached.size as cached_size, cached.sha256 as cached_sha256, cached.md5 as cached_md5
+
+            FROM fs LEFT JOIN cached USING (path)
 
             WHERE fs.path LIKE :path_like AND
-                (fs.mtime != cached.mtime OR cached.path IS NULL)
+                (cached.path IS NULL OR
+                    (fs.sha256 AND fs.sha256 != cached.sha256)
+                    OR
+                    (fs.md5 AND fs.md5 != cached.md5)
+                    OR
+                    -- incorrect way to check (hash is NULL or '')
+                    (NOT fs.sha256 AND NOT fs.md5 AND (fs.mtime != cached.mtime))
+                    -- original mtime check - could be fs.mtime <= cached.mtime
+                    OR (fs.mtime != cached.mtime OR cached.path IS NULL)
+                )
```

Comment:
@barabo this isn't quite right. It might make sense to do the mtime comparison in Python by filtering the results as they come through. We want to use mtime comparisons to make sure the upstream stage has the correct sums, and then make sure 'indexed' aka cached matches the sums. But we didn't want to wait to actually recompute all checksums on conda-forge.

Reply:
however `changed_packages` doesn't reset to the empty list `[]` after re-indexing